guikit/threads.py

```diff
@@ -196,17 +217,51 @@ def abort_thread(self, ident: int) -> None:
         Raises:
             KeyError: If the thread identifier is not in the ThreadPool
         """
-        try:
-            self._workers[ident].abort = True
-        except KeyError as key_err:
-            key_err_ = KeyError(f"Thread with index: {ident} is not in the ThreadPool.")
-            logger.exception(key_err_)
-            raise key_err_ from key_err
+        self.get_thread(ident).abort = True
+
+    def join_thread(self, ident: int) -> None:
+        """Join the thread specified by `ident`.
+
+        Args:
+            ident: Thread identifier
+
+        Raises:
+            KeyError: If the thread identifier is not in the ThreadPool
+        """
+        self.get_thread(ident).join()
 
     def post_event(self, event: ThreadResult):
         """Adds an event to the event loop of the main thread."""
         wx.PostEvent(self._window, event)
 
+    def stop_threads(self):
+        """Stop all worker threads and wait for them to finish.
+
+        Todo:
+            Leave daemon threads running?
+            Race condition if a thread is created while this function is running
+        """
+        # Copy the list of worker threads in case the running threads try to
+        # modify self._workers
+        with self._workers_lock:
+            workers = self._workers.values()
+
+        # Tell workers that we want to abort
+        for worker in workers:
+            worker.abort = True
+
+        # Wait for all workers to finish
+        for worker in workers:
+            worker.join()
```

Comment:
What happens if a worker has finished after setting it to `abort` if we try to `join` it?

Reply:
That's ok too! The ``join()`` will just return instantly.